File: Leetcode/cpp/sol102.cpp

// https://leetcode.com/problems/binary-tree-level-order-traversal/submissions/1661789876/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> res;
    void helper(TreeNode* current, int level){
        if(!current){
            return;
        }
        if(level == res.size()){
            res.push_back({});
        }
        res[level].push_back(current->val);
        helper(current->left, level+1);
        helper(current->right, level+1);
    }

    vector<vector<int>> levelOrder(TreeNode* root) {
        helper(root, 0);
        return res;
    }
};

File: Leetcode/typescript/sol102.ts

//https://leetcode.com/problems/binary-tree-level-order-traversal/description/

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function levelOrder(root: TreeNode | null): number[][] {
    if(!root){
        return [];
    }
    const Q:[TreeNode | null, number][] = [[root, 0]];
    const mapa = new Map();

    while(Q.length > 0){
        const popped: [TreeNode | null, number] | undefined = Q.shift();
        if(!popped){
            continue;
        }

        const [node, level] = popped;
        if(node != null){
            console.log(node.val, level)
            if(!(level in mapa)){
                mapa[level] = [];
            }
            mapa[level].push(node.val)

            if(node.left){
                Q.push([node.left, level + 1])
            }
            if(node.right){
                Q.push([node.right, level + 1])
            }
        }
    }

    return Object.values(mapa);
};